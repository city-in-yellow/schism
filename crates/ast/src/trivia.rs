#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct Range {
    pub start: usize,
    pub end: usize,
}

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Trivia<T> {
    pub inner: T,
    pub span: Range,
}

pub fn new<T>(inner: T, span: Range) -> Trivia<T> {
    Trivia { inner, span }
}

pub trait WithTrivia {
    fn pretty_string(&self, indent: usize) -> impl AsRef<str>;
}
