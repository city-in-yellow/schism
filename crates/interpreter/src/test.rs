use ast::trivia::WithTrivia;
use indoc::indoc;
use insta::assert_snapshot;

use crate::run_on_src;

fn e(s: &str) -> String {
    match run_on_src(s) {
        Ok(o) => {
            format!("{}\n", o.pretty_string(0))
        }
        Err(e) => e.to_report(s),
    }
}

#[test]
fn test() {
    assert_snapshot!(e("1"));
    assert_snapshot!(e("()"));
    assert_snapshot!(e(r#""hi there""#));
    assert_snapshot!(e(r#"foo is 1. foo"#));
    assert_snapshot!(e(r#"[1, 2, 3]"#));
    assert_snapshot!(e(r#"foo bar is bar. foo 1"#));
    assert_snapshot!(e(r#"1 + 2"#));
    assert_snapshot!(e(indoc! {r#"
        foo bar baz is
            barbaz is bar + baz.
            bazbar is bar * baz.
            bazbar - barbaz.
        
        foo 3 5
    "#}));
    assert_snapshot!(e(r#"[1, 2, 3] + 12"#));
    assert_snapshot!(e(r#"1 + "2""#));
    assert_snapshot!(e(indoc! {r#"
        foo bar baz bil is
            barbaz is bar + baz.
            barbaz + bil.
        
        foo 3 5 "a"
    "#}));
}
