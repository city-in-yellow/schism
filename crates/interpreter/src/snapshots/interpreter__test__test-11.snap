---
source: crates/interpreter/src/test.rs
expression: "e(indoc! {\n            r#\"\n        foo bar baz bil is\n            barbaz is bar + baz.\n            barbaz + bil.\n        \n        foo 3 5 \"a\"\n    \"#\n        })"
---
ꡃꡨꡀꡃꡕ: kaathe.encyc.sim.pat_err
   ╭─[<unknown>:1:1]
   │
 1 │ ╭─▶ foo bar baz bil is
   ┆ ┆   
 5 │ ├─▶ foo 3 5 "a"
   │ │           ─┬─   
   │ ╰───────────────── 「argument type is invalid」
   │              │    
   │              ╰──── expected one of (Number, Decimal, List) but this is type String
───╯

